### DISCUSSION TITLE
How to maintain transient state in client component?

### DISCUSSION THREAD

**Discussion by vedantroy** [vercel/next.js#50777] [Help] [answered]
### Summary

I have a select element, where, whenever the user clicks on an option, it submits an action to the server.

The issue is--whenever I select a new option, the form runs the server action, which causes the client component to re-render from scratch. This causes the component to reset its active value.

Is there a way I can fix this?

### Additional information

```js
"use client"
import React, { useState, useRef } from 'react';
import { changeTable } from './actions';

const SelectForm = ({ options }) => {
  const [selectedOption, setSelectedOption] = useState(options[0]);
  const formRef = useRef<HTMLFormElement>(null);

  const handleChange = event => {
    setSelectedOption(event.target.value);
    
    if (formRef.current) {
      formRef.current.submit();
    }
  };

  return (
    <form ref={formRef} action={changeTable}>
      <div className="inline-block relative w-64">
        <select
          value={selectedOption}
          onChange={handleChange}
          className="block appearance-none w-full bg-white border border-gray-400 hover:border-gray-500 px-4 py-2 pr-8 rounded shadow leading-tight focus:outline-none focus:shadow-outline"
        >
          {options.map(option => (
            <option key={option} value={option}>
              {option}
            </option>
          ))}
        </select>
        <div className="pointer-events-none absolute inset-y-0 right-0 flex items-center px-2 text-gray-700">
          <svg className="fill-current h-4 w-4" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20"><path d="M10 12l-6-6 1.414-1.414L10 9.172l4.586-4.586L16 6z"/></svg>
        </div>
      </div>
    </form>
  );
};

export default SelectForm;
```

**Comment by vedantroy** [comment]
My solution:
```
"use client"
import React, { useState, useRef } from 'react';
import { changeTable } from './actions';

const SelectForm = ({ options }) => {
  const [selectedOption, setSelectedOption] = useState(options[0]);
  const formRef = useRef<HTMLFormElement>(null);

  const handleChange = event => {
    setSelectedOption(event.target.value);
    
    if (formRef.current) {
      formRef.current.submit();
    }
  };

  return (
    <form ref={formRef} action={async formData => {
        await changeTable(formData)
        setSelectedOption(formData.get("tableName"))
    }}>
      <div className="inline-block relative w-64">
        <select
          value={selectedOption}
          onChange={handleChange}
          name="tableName"
          className="block appearance-none w-full bg-white border border-gray-400 hover:border-gray-500 px-4 py-2 pr-8 rounded shadow leading-tight focus:outline-none focus:shadow-outline"
        >
          {options.map(option => (
            <option key={option} value={option}>
              {option}
            </option>
          ))}
        </select>
        <div className="pointer-events-none absolute inset-y-0 right-0 flex items-center px-2 text-gray-700">
          <svg className="fill-current h-4 w-4" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20"><path d="M10 12l-6-6 1.414-1.414L10 9.172l4.586-4.586L16 6z"/></svg>
        </div>
      </div>
    </form>
  );
};

export default SelectForm;
```

Leaving this open in-case someone finds a better solution.

**Reply by icyJoseph** [reply, marked as answer]
This does't really seem to be correct though, it might be later since this is all in alpha stage, but I see an error in the console, every time I submit. 

```console
VM205:1 Uncaught Error: A React form was unexpectedly submitted.
    at <anonymous>:1:7
(anonymous) @ VM205:1
VM206:1 Uncaught SyntaxError: missing ) after argument list
```

Anyway, may I ask one thing though, why do this at all? The action prop, in forms with server actions, is to enable submit without JS. That is to enable a fully functional server components form. Yours is a client component, it necessarily requires JS to operate, so you could just something like below. 

> Note that I am not saying that server actions and client components do not mix, they do! However in this case it feels like we are bending over to achieve something that could be easily done in another way. Then again, this is alpha tech, so lingering more than 10 minutes on these API problems feels a bit unnecessary

```tsx
"use client";
import {
  type ChangeEvent,
  //   useTransition,
  useState,
  startTransition,
  useId,
} from "react";
import { changeTable } from "./actions";

const SelectForm = ({ options }: { options: string[] }) => {
  const [selectedOption, setSelectedOption] = useState(options[0]);
  const formId = useId();

  const handleChange = (event: ChangeEvent<HTMLSelectElement>) => {
    setSelectedOption(event.target.value);

    const currentForm = event.target.form;

    if (!currentForm) return;

    startTransition(() => {
      changeTable(new FormData(currentForm));
    });
  };

  return (
    <form id={formId}>
      <div className="inline-block relative w-64">
        <select
          value={selectedOption}
          onChange={handleChange}
          className="block appearance-none w-full text-black bg-white border border-gray-400 hover:border-gray-500 px-4 py-2 pr-8 rounded shadow leading-tight focus:outline-none focus:shadow-outline"
          form={formId}
          name="select-option"
        >
          {options.map((option) => (
            <option key={option} value={option}>
              {option}
            </option>
          ))}
        </select>
        <div className="pointer-events-none absolute inset-y-0 right-0 flex items-center px-2 text-gray-700">
          <svg
            className="fill-current h-4 w-4"
            xmlns="http://www.w3.org/2000/svg"
            viewBox="0 0 20 20"
          >
            <path d="M10 12l-6-6 1.414-1.414L10 9.172l4.586-4.586L16 6z" />
          </svg>
        </div>
      </div>
    </form>
  );
};

export default SelectForm;
```

And this is sort of documented here:

> You can invoke Server Actions using the following methods:
> 
> - Using action: React's action prop allows invoking a Server Action on a `<form>` element.
> - Using formAction: React's formAction prop allows handling `<button>`, `<input type="submit">`, and `<input type="image">` elements in a `<form>`.
> - Custom Invocation with startTransition: Invoke Server Actions without using action or formAction by using startTransition. This method disables Progressive Enhancement

**Comment by devjiwonchoi** [comment]
Seems like you're not actually using the feature `Server Actions` which is in the `alpha` stage at the moment.

Also, your code

```js
if (formRef.current) {
      formRef.current.submit();
    }
```

submits every time the option is selected, which is not a great option from a UX perspective.
I would recommend you add a `Submit Button` and manage the `submit` request.

If you do need to submit it every time the option selection, I would recommend you save it to a cookie.

You can do this in various ways but since you've stated, you can also do this by using the `Server Actions`.
Note that you cannot use the `Server Actions` directly inside the Client Component, but you can import it.

> Server Actions cannot be defined within Client Components, but they can be imported.

Also, note that while importing the `Server Actions` Component, `'use server'` needs to be declared on the top of the file.

> To use Server Actions in Client Components, you can import the action from a file containing a top-level "use server" directive.

In summary, add a `Submit Button` to handle your submit request. If not, save it in the cookie.

## Reference
- Server Actions: https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions
- Server Actions with Client Components: https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions#usage-with-client-components
